File: dotnet/Common/AzureOpenAIClientFactory.cs
using Azure;
using Azure.AI.OpenAI;
using Azure.Identity;
using Microsoft.Extensions.AI;
using Microsoft.Extensions.Configuration;

namespace WorkflowLab.Common;

/// <summary>
/// Factory class for creating Azure OpenAI chat clients with multiple authentication options.
/// Supports configuration from appsettings.json and environment variables.
/// </summary>
public static class AzureOpenAIClientFactory
{
    private static IConfiguration? _configuration;
    private static string? _configPath;

    /// <summary>
    /// Gets the configuration, loading from appsettings.Local.json in the dotnet folder and environment variables.
    /// </summary>
    private static IConfiguration Configuration
    {
        get
        {
            if (_configuration == null)
            {
                _configPath = FindConfigPath(AppContext.BaseDirectory);
                _configuration = new ConfigurationBuilder()
                    .SetBasePath(_configPath)
                    .AddJsonFile("appsettings.Local.json", optional: true, reloadOnChange: true)
                    .AddEnvironmentVariables()
                    .Build();
            }
            return _configuration;
        }
    }

    /// <summary>
    /// Finds the dotnet folder by traversing up the directory tree.
    /// </summary>
    private static string FindConfigPath(string startPath)
    {
        var currentDir = new DirectoryInfo(startPath);

        // Traverse up to find the 'dotnet' folder
        while (currentDir != null)
        {
            if (currentDir.Name.Equals("dotnet", StringComparison.OrdinalIgnoreCase))
            {
                return currentDir.FullName;
            }
            currentDir = currentDir.Parent;
        }

        // Fallback to start path if dotnet folder not found
        return startPath;
    }

    /// <summary>
    /// Creates an Azure OpenAI chat client with support for multiple authentication methods:
    /// 1. API Key authentication (AzureOpenAI:ApiKey or AZURE_OPENAI_API_KEY)
    /// 2. Service Principal authentication (TenantId, ClientId, ClientSecret)
    /// 3. Managed Identity / DefaultAzureCredential (fallback)
    /// 
    /// Configuration can come from appsettings.json or environment variables.
    /// Environment variables take precedence over appsettings.json.
    /// </summary>
    public static IChatClient CreateChatClient()
    {
        // Force configuration loading to set _configPath
        _ = Configuration;
        
        // Display config location
        Console.WriteLine($"📁 Config path: {_configPath}");
        Console.WriteLine($"📄 Config file: {Path.Combine(_configPath ?? "", "appsettings.Local.json")}");
        Console.WriteLine();

        // Get endpoint (required)
        var endpoint = GetConfigValue("AzureOpenAI:Endpoint", "AZURE_OPENAI_ENDPOINT")
            ?? throw new InvalidOperationException(
                "Azure OpenAI endpoint is not configured. " +
                "Set 'AzureOpenAI:Endpoint' in appsettings.json or 'AZURE_OPENAI_ENDPOINT' environment variable.");

        // Get deployment name (optional, default: gpt-4o-mini)
        // Check in order: AZURE_OPENAI_DEPLOYMENT_NAME, AZURE_AI_MODEL_DEPLOYMENT_NAME, AzureOpenAI:DeploymentName
        var deploymentName = Configuration["AZURE_OPENAI_DEPLOYMENT_NAME"] 
            ?? Configuration["AZURE_AI_MODEL_DEPLOYMENT_NAME"]
            ?? Configuration["AzureOpenAI:DeploymentName"]
            ?? "gpt-4o-mini";

        Console.WriteLine($"✅ Configuration loaded");
        Console.WriteLine($"   Endpoint: {endpoint}");
        Console.WriteLine($"   Deployment: {deploymentName}");
        Console.WriteLine();

        // Option 1: API Key authentication
        var apiKey = GetConfigValue("AzureOpenAI:ApiKey", "AZURE_OPENAI_API_KEY");
        if (!string.IsNullOrEmpty(apiKey))
        {
            Console.WriteLine("Using API Key authentication");
            return new AzureOpenAIClient(new Uri(endpoint), new AzureKeyCredential(apiKey))
                .GetChatClient(deploymentName)
                .AsIChatClient();
        }

        // Option 2: Service Principal authentication (Tenant ID, Client ID, Client Secret)
        var tenantId = GetConfigValue("AzureOpenAI:TenantId", "AZURE_TENANT_ID");
        var clientId = GetConfigValue("AzureOpenAI:ClientId", "AZURE_CLIENT_ID");
        var clientSecret = GetConfigValue("AzureOpenAI:ClientSecret", "AZURE_CLIENT_SECRET");

        if (!string.IsNullOrEmpty(tenantId) && !string.IsNullOrEmpty(clientId) && !string.IsNullOrEmpty(clientSecret))
        {
            Console.WriteLine("Using Service Principal authentication");
            var credential = new ClientSecretCredential(tenantId, clientId, clientSecret);
            return new AzureOpenAIClient(new Uri(endpoint), credential)
                .GetChatClient(deploymentName)
                .AsIChatClient();
        }

        // Option 3: Managed Identity / DefaultAzureCredential (fallback)
        Console.WriteLine("Using Managed Identity / DefaultAzureCredential authentication");
        return new AzureOpenAIClient(new Uri(endpoint), new DefaultAzureCredential())
            .GetChatClient(deploymentName)
            .AsIChatClient();
    }

    /// <summary>
    /// Gets a configuration value, checking environment variable first, then config keys.
    /// Environment variables take precedence if both are set.
    /// </summary>
    private static string? GetConfigValue(string appSettingsKey, string environmentVariable)
    {
        // Check environment variable first (highest precedence)
        var envValue = Environment.GetEnvironmentVariable(environmentVariable);
        if (!string.IsNullOrEmpty(envValue))
        {
            return envValue;
        }

        // Check config file with environment variable key (e.g., AZURE_OPENAI_ENDPOINT)
        var configEnvValue = Configuration[environmentVariable];
        if (!string.IsNullOrEmpty(configEnvValue))
        {
            return configEnvValue;
        }

        // Fall back to nested config key (e.g., AzureOpenAI:Endpoint)
        return Configuration[appSettingsKey];
    }
}

File: python/common/azure_openai_client_factory.py
"""
Azure OpenAI Client Factory

Factory class for creating Azure OpenAI chat clients with multiple authentication options.
Supports configuration from environment variables.
"""

import os
from openai import AzureOpenAI
from azure.identity import AzureCliCredential, ClientSecretCredential, DefaultAzureCredential


def create_chat_client() -> AzureOpenAI:
    """
    Creates an Azure OpenAI chat client with support for multiple authentication methods:
    1. API Key authentication (AZURE_OPENAI_API_KEY)
    2. Service Principal authentication (AZURE_TENANT_ID, AZURE_CLIENT_ID, AZURE_CLIENT_SECRET)
    3. Azure CLI credential
    4. Managed Identity / DefaultAzureCredential (fallback)
    
    Returns:
        AzureOpenAI: Configured Azure OpenAI client
        
    Raises:
        ValueError: If required configuration is missing
    """
    # Get endpoint (required) - check both possible env var names
    endpoint = os.environ.get("AZURE_OPENAI_ENDPOINT")
    if not endpoint:
        raise ValueError(
            "Azure OpenAI endpoint is not configured. "
            "Set 'AZURE_OPENAI_ENDPOINT' environment variable."
        )
    
    # Get deployment name - check both possible env var names
    deployment = os.environ.get("AZURE_OPENAI_DEPLOYMENT_NAME") or \
                 os.environ.get("AZURE_AI_MODEL_DEPLOYMENT_NAME") or \
                 "gpt-4o-mini"
    api_version = "2024-02-15-preview"
    
    print(f"Using endpoint: {endpoint}")
    print(f"Using deployment: {deployment}")
    
    # Option 1: API Key authentication
    api_key = os.environ.get("AZURE_OPENAI_API_KEY")
    if api_key:
        print("Using API Key authentication")
        return AzureOpenAI(
            azure_endpoint=endpoint,
            api_key=api_key,
            api_version=api_version
        )
    
    # Option 2: Service Principal authentication
    tenant_id = os.environ.get("AZURE_TENANT_ID")
    client_id = os.environ.get("AZURE_CLIENT_ID")
    client_secret = os.environ.get("AZURE_CLIENT_SECRET")
    
    if tenant_id and client_id and client_secret:
        print("Using Service Principal authentication")
        credential = ClientSecretCredential(
            tenant_id=tenant_id,
            client_id=client_id,
            client_secret=client_secret
        )
        token = credential.get_token("https://cognitiveservices.azure.com/.default")
        return AzureOpenAI(
            azure_endpoint=endpoint,
            api_key=token.token,
            api_version=api_version
        )
    
    # Option 3: Try Azure CLI credential first
    try:
        print("Trying Azure CLI authentication...")
        credential = AzureCliCredential()
        token = credential.get_token("https://cognitiveservices.azure.com/.default")
        print("Using Azure CLI authentication")
        return AzureOpenAI(
            azure_endpoint=endpoint,
            api_key=token.token,
            api_version=api_version
        )
    except Exception:
        pass
    
    # Option 4: Fallback to DefaultAzureCredential (Managed Identity, etc.)
    print("Using Managed Identity / DefaultAzureCredential authentication")
    credential = DefaultAzureCredential()
    token = credential.get_token("https://cognitiveservices.azure.com/.default")
    return AzureOpenAI(
        azure_endpoint=endpoint,
        api_key=token.token,
        api_version=api_version
    )


def get_deployment_name() -> str:
    """Get the deployment name from environment or default."""
    return os.environ.get("AZURE_OPENAI_DEPLOYMENT_NAME") or \
           os.environ.get("AZURE_AI_MODEL_DEPLOYMENT_NAME") or \
           "gpt-4o-mini"
